python/lsst/obs/base/ingest.py

```diff
@@ -318,16 +365,181 @@ def _calculate_dataset_info(self, header, filename):
+                                     file_in_dir, e)
+                    bad_index_files.add(possible_index)
+                else:
+                    good_index_files.add(possible_index)
+                    self.log.debug("Extracted potential index metadata from inferred file %s", possible_index)
+
+                for file_in_dir in files_in_directory:
+                    if file_in_dir in index:
+                        file = os.path.abspath(os.path.join(directory, file_in_dir))
+                        if file in index_entries:
+                            # ObservationInfo overrides raw metadata
+                            if isinstance(index[file_in_dir], ObservationInfo) \
+                                    and not isinstance(index_entries[file], ObservationInfo):
+                                self.log.warning("File %s already specified in an index file but overriding"
+                                                 " with ObservationInfo content from %s",
+                                                 file, possible_index)
+                            else:
+                                self.log.warning("File %s already specified in an index file, "
+                                                 "ignoring content from %s", file, possible_index)
+                        else:
+                            index_entries[file] = index[file_in_dir]
+
+        # Remove files from list that have index entries and also
+        # any files that we determined to be explicit index files
+        # or any index files that we failed to read.
+        files = files_set - set(index_entries) - good_index_files - bad_index_files
+
+        return index_entries, files, good_index_files, bad_index_files
+
+    def processIndexEntries(self, index_entries):
+        """Convert index entries to RawFileData.
+
+        Parameters
+        ----------
+        index_entries : `dict[str, Any]`
+            Dict indexed by name of file to ingest and with keys either
+            raw metadata or translated `ObservationInfo`.
+
+        Returns
+        -------
+        data :  `RawFileData`
+            A structure containing the metadata extracted from the file,
+            as well as the original filename.  All fields will be populated,
+            but the `RawFileData.dataId` attribute will be a minimal
+            (unexpanded) `DataCoordinate` instance.
+        """
+        fileData = []
+        for filename, metadata in index_entries.items():
+            datasets = [self._calculate_dataset_info(metadata, filename)]
+            instrument, formatterClass = self._determine_instrument_formatter(datasets[0].dataId, filename)
+            if instrument is None:
+                datasets = []
+            fileData.append(RawFileData(datasets=datasets, filename=filename,
+                                        FormatterClass=formatterClass, instrumentClass=instrument))
```

Comment:
If `instrument is None` and `datasets == []`, this is appending essentially no useful information? What would the caller of this method do with a no-instrument, no-dataset, base-formatter file?

Reply:
This is the standard sentinel for indicating to the caller that something went wrong in populating the RawFileData. As I write this I realize I was not using this information in the caller so will add a matching log message. I am not logging in processIndexEntries in case it becomes vectorized at some point.